File: dsa/Tree/inorder.java
// repeat step 2-4 while tree!= null
//      inorder(tree->left)
//      inorder(tree->data)
//      inorder(tree->right)
// end

import java.util.Scanner;

class Node {
    int data;
    Node left, right;

    public Node(int item) {
        data = item;
        left = right = null;
    }
}

class BinarySearchTree {
    Node root;

    // Logic: Standard BST Insertion
    Node insert(Node root, int data) {
        if (root == null) {
            root = new Node(data);
            return root;
        }
        if (data < root.data)
            root.left = insert(root.left, data);
        else if (data > root.data)
            root.right = insert(root.right, data);
        
        return root;
    }

    // ALGO: Inorder (Left -> Data -> Right)
    void inorder(Node tree) {
        if (tree != null) {
            inorder(tree.left);
            System.out.print(tree.data + " ");
            inorder(tree.right);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BinarySearchTree bst = new BinarySearchTree();

        System.out.print("Enter number of nodes: ");
        int n = sc.nextInt();

        System.out.println("Enter " + n + " values to build the BST:");
        for (int i = 0; i < n; i++) {
            int val = sc.nextInt();
            bst.root = bst.insert(bst.root, val);
        }

        System.out.println("\nInorder traversal result (Sorted Order):");
        bst.inorder(bst.root);
        
        sc.close();
    }
}

File: dsa/Tree/inorder.c
#include <stdio.h>
#include <stdlib.h>

// Structure for the BST node
struct node {
    int data;
    struct node *left;
    struct node *right;
};

// Function to create a new node
struct node* newNode(int val) {
    struct node* temp = (struct node*)malloc(sizeof(struct node));
    temp->data = val;
    temp->left = temp->right = NULL;
    return temp;
}

// ALGO: Inorder Traversal (Left, Root, Right)
void inorder(struct node* tree) {
    if (tree != NULL) {
        inorder(tree->left);       // Step 2: Visit Left Subtree
        printf("%d ", tree->data); // Step 3: Print Data (Root)
        inorder(tree->right);      // Step 4: Visit Right Subtree
    }
}

int main() {
    /* Creating a sample BST:
          50
         /  \
        30   70
    */
    struct node* root = newNode(50);
    root->left = newNode(30);
    root->right = newNode(70);

    printf("Inorder Traversal Result: ");
    inorder(root);

    return 0;
}